machine/machine_core/ssh_connection.py

```diff
@@ -311,7 +311,7 @@ def execute(self, command=None, script=None, input=None, environment={},
                 # use shlex.join() once Python 3.8 is available everywhere
                 command = ' '.join(shlex.quote(arg) for arg in command)
 
-            cmd += [command]
+            cmd += ['set -e; ' + command]
```

Comment:
SSH is so very evil...

Reply:
It is, but do you have a better idea? Given how much crap and brokenness that I found recently with this, I am more convinced than ever that `set -e` by default is the right thing to do. Of course the "how" looks quite horrible..

Reply:
I was actually thinking to add `'set -x;`' as a separate commandline argument to ssh, with the knowledge that SSH is going to paste everything together again anyway (giving you the same result).

This would also allow you to move the `set -e` above the computation of the command itself (ie: the list-flattening stuff) and replace that with `.append()` or `.extend()` in order to avoid the extra variable assignment (and remove the comment about `shlex.join()`.

Reply:
Something like:

```python
cmd.append('set -e;')
if not isinstance(command, str):
    cmd.extend(shlex.quote(arg) for arg in command)
else:
    cmd.append(str)
````

Reply:
But that's going to need the if/else case for the logging again. It really is the simplest way to just unify the two cases (list/string) very early IMHO. I can't find any iteration of that which is significantly simpler, so I'm just going to drop that commit. It's not necessary for what I want to achieve after all, and we can banter about this in a less intrusive PR (in terms of how many tests it needs).

Reply:
> I was actually thinking to add 'set -e;' as a separate commandline argument to ssh

That's a great idea! Done.

Reply:
I'd be happy enough to merge this already (since we bikeshed enough) but your changes after this message seem not to have been pushed yet, so I'll wait.

Reply:
Argh! I pushed it to my fork instead of origin last night. There go 45 green tests..